### DISCUSSION TITLE
Spring Event로 Member 삭제 시 Post, Trip 삭제하기

### DISCUSSION THREAD

**Discussion by Combi153** [woowacourse-teams/2023-trip-draw#359] [기록]
### 요구사항

- Member 삭제 시 Member의 data 삭제
- 삭제 시 DataIntegrityViolationException 를 피하기 위해 삭제 순서를 고려해야 함
    1. Post 삭제
    2. Trip 삭제
    3. Member 삭제

### 삭제 순서

@Order 애노테이션 사용

> **Ordering Listeners**
It is also possible to define the order in which listeners for a certain event are to be invoked. To do so, add Spring's common `@Order` annotation alongside this event listener annotation.
> 

[EventListener (Spring Framework 6.0.11 API)](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/event/EventListener.html)

[Order (Spring Framework 6.0.11 API)](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/annotation/Order.html)

[How can I ensure that a Spring @EventListener will be invoked first?](https://stackoverflow.com/questions/35605505/how-can-i-ensure-that-a-spring-eventlistener-will-be-invoked-first)

### 코드 구현 미리보기
```java
@Service
public class MemberService {

    (...)

    @Transactional
    public void deleteByCode(String code) {
        Long memberId = Long.valueOf(jwtTokenProvider.extractAccessToken(code));
        publisher.publishEvent(new DeleteMemberEvent(memberId));
        memberRepository.deleteById(memberId);
    }
}
```

```java
@Component
public class PostDeleteEventHandler {

    private final PostRepository postRepository;

    public PostDeleteEventHandler(PostRepository postRepository) {
        this.postRepository = postRepository;
    }

    @EventListener
    public void deletePostByMemberId(DeleteMemberEvent event) {
        postRepository.deleteByMemberId(event.memberId());
    }
}
```

### Transaction
성공 예시

- deleteByCode 메서드 트랜잭션 시작
    - 이벤트 발급
    - PostDeleteEventHandler 이벤트 처리 : Post 삭제
    - TripDeleteEventHandler 이벤트 처리 : Trip 삭제
    - memberRepository Member 삭제

실패 예시

- deleteByCode 메서드 트랜잭션 시작
    - 이벤트 발급
    - PostDeleteEventHandler 이벤트 처리 : Post 삭제
    - TripDeleteEventHandler 이벤트 처리 : Trip 삭제
    - memberRepository Member 삭제 → 예외 발생
    - 트랜잭션 rollback
        - Post, Trip, Member 모든 데이터 rollback

### Test
[Record Spring Events When Testing Spring Boot Applications - rieckpil](https://rieckpil.de/record-spring-events-when-testing-spring-boot-applications/)

### 참고자료
[Transaction-bound Events :: Spring Framework](https://docs.spring.io/spring-framework/reference/data-access/transaction/event.html)

**Comment by greeng00se** [comment]
깔끔하게 정리되어 있어 해당 PR을 이해하기 더욱 편하네요 👍
